src/server/master_server/chunk_server_heartbeat_monitor_task.cc

```diff
@@ -0,0 +1,177 @@
+#include "src/server/master_server/chunk_server_heartbeat_monitor_task.h"
+
+#include <chrono>
+#include <memory>
+
+#include "absl/strings/str_cat.h"
+#include "src/common/protocol_client/chunk_server_control_service_client.h"
+#include "src/common/system_logger.h"
+#include "src/server/master_server/chunk_server_manager.h"
+
+using gfs::common::ConfigManager;
+using gfs::service::ChunkServerControlServiceClient;
+using google::protobuf::util::Status;
+using google::protobuf::util::StatusOr;
+using protos::grpc::CheckHeartBeatReply;
+using protos::grpc::CheckHeartBeatRequest;
+
+namespace gfs {
+
+namespace server {
+
+ChunkServerHeartBeatMonitorTask::ChunkServerHeartBeatMonitorTask()
+    : thread_(nullptr),
+      terminate_promise_(nullptr),
+      terminate_future_(nullptr),
+      config_mgr_(nullptr) {}
+
+ChunkServerHeartBeatMonitorTask::~ChunkServerHeartBeatMonitorTask() {
+  Terminate();
+}
+
+ChunkServerHeartBeatMonitorTask*
+ChunkServerHeartBeatMonitorTask::GetInstance() {
+  static ChunkServerHeartBeatMonitorTask instance;
+
+  return &instance;
+}
+
+void ChunkServerHeartBeatMonitorTask::Start(const std::string& config_file) {
+  if (!this->terminate_promise_) {
+    LOG(INFO) << "Chunk server heartbeat monitor task is starting...";
+
+    // Instantiate a ConfigManager with the given filename
+    StatusOr<ConfigManager*> config_manager_or(
+        ConfigManager::GetConfig(config_file));
+
+    this->config_mgr_ =
+        std::unique_ptr<ConfigManager>(config_manager_or.ValueOrDie());
+
+    // Create the promise used to signal the background thread to terminate.
+    this->terminate_promise_ =
+        std::unique_ptr<std::promise<void>>(new std::promise<void>());
+
+    this->terminate_future_ =
+        std::unique_ptr<std::future<void>>(new std::future<void>(
+            std::move(this->terminate_promise_->get_future())));
+
+    // Create the background thread to run the heartbeat monitoring work.
+    this->thread_ = std::unique_ptr<std::thread>(new std::thread(
+        &ChunkServerHeartBeatMonitorTask::MonitorHeartBeat, this));
+  }
+}
+
+void ChunkServerHeartBeatMonitorTask::MonitorHeartBeat() {
+  LOG(INFO) << "Chunk server heartbeat monitor task is now running in the "
+               "background...";
+
+  // TODO(bmokutub): Make this configurable, with the config manager.
+  // Max number of times we should attempt to send heartbeat consecutively to a
+  // chunkserver, before declaring it as unavailable.
+  const ushort max_attempts = 3;
+
+  while (true) {
+    for (auto& chunk_server :
+         ChunkServerManager::GetInstance().chunk_servers_map_) {
+      // Check if we have been asked to terminate before proceeding.
+      if (IsTerminateSignalled()) {
+        return;
+      }
+
+      CheckHeartBeatRequest request;
+
+      auto server_address =
+          absl::StrCat(this->config_mgr_->ResolveHostname(
```

Comment:
Should resolve host name be only relevant when resolve_hostname is true? In this vein, I think the constructor param needs to take resolve_hostname in the main function. I have encountered similar issue in chunk server manager's service and fixed with passing (config, resolve_hostname) and only resolve hosname when resolve_hostname is true.

Reply:
now passing resolve_hostname